### PR TITLE
Enable Cutover callback in config for copydb

### PR DESCRIPTION
**Pull request by floriecai**
Refactoring the code so the shared config between copydb and sharding allow `CutoverLock` and `CutoverUnlock` callbacks.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -618,10 +618,23 @@ type Config struct {
...
 
+	// Report when ghostferry is entering cutover
+	CutoverLock   			HTTPCallback
+
+	// Report when ghostferry is finished cutover
+	CutoverUnlock 			HTTPCallback
+
+	// These two values configure the amount of times Ferry should attempt to
+	// retry acquiring the cutover lock, and for how long the Ferry should wait
+	// before attempting another lock acquisition
```
Comment (shuhaowu) [inline config.go:634]: Probably should add:
MaxCutoverRetries Default = 1
CutoverRetryWaitSeconds Second Default ...

File: config.go [hunk 2]
```diff
@@ -618,10 +618,23 @@ type Config struct {
...
 	ErrorCallback HTTPCallback
 
+	// Report when ghostferry is entering cutover
+	CutoverLock   			HTTPCallback
+
+	// Report when ghostferry is finished cutover
+	CutoverUnlock 			HTTPCallback
+
+	// These two values configure the amount of times Ferry should attempt to
+	// retry acquiring the cutover lock, and for how long the Ferry should wait
```
Comment (shuhaowu) [inline config.go:633]: Probably should add before:

If the server for CutoverLock returns a non OK status, then these two values ... <carry on>

File: copydb/cmd/main.go [hunk 3]
```diff
@@ -121,7 +120,7 @@ func main() {
 		return
 	}
 
-	if resumeState == nil {
+	if config.StateToResumeFrom == nil {
```
Comment (floriecai) [inline copydb/cmd/main.go:123]: `resumeState` was previously only set if we passed in a path for the statedump file. We want to be able to pass in the state dump directly in the config through the field `StateToResumeFrom` as well.

File: copydb/copydb.go [hunk 4]
```diff
@@ -100,6 +100,7 @@ func (this *CopydbFerry) Run() {
 	// binlog streamer catching up.
 	this.Ferry.WaitUntilBinlogStreamerCatchesUp()
 
+	cutoverStart := this.Ferry.StartCutover()
```
Comment (shuhaowu) [inline copydb/copydb.go:103]: Can you comment on what this function is supposed to do? Like:


> Optionally (configurable) post to an HTTP endpoint telling that service Ghostferry is ready for cutover. The external service can then perform steps needed immediately prior to cutover. For example, an external service can set database to be readonly when the this endpoint is invoked.

File: copydb/copydb.go [hunk 5]
```diff
@@ -112,6 +113,7 @@ func (this *CopydbFerry) Run() {
 
 	this.Ferry.StopTargetVerifier()
 
+	this.Ferry.EndCutover(cutoverStart)
```
Comment (shuhaowu) [inline copydb/copydb.go:116]: Similar, I would add a comment. This flow is already a bit confusing, adding more without a comment will likely add to confusion.

**Review by shuhaowu — APPROVED** [review]
Minor comment changes requested, other than that this looks good.

Not relevant for now, but for the future: we want to consider rewording some of the steps in cutover (and thus the function and variable names), as the list is long and confusing. For example, if `AutomaticCutover` is false, Ghostferry does not notify  via callback that copy is complete and cutover can occur anytime. The current `cutover lock` callback occurs immediately prior to flushing the binlog streamer and cutting over, which may be too late for certain applications.

File: copydb/cmd/main.go [hunk 6]
```diff
@@ -99,6 +78,26 @@ func main() {
...
+		}
+
+		resumeState := &ghostferry.SerializableState{}
+		parser := json.NewDecoder(f)
+		err = parser.Decode(&resumeState)
+		if err != nil {
+			errorAndExit(fmt.Sprintf("failed to parse state file: %v", err))
+		}
+
+		config.StateToResumeFrom = resumeState
```
Comment (shuhaowu) [inline copydb/cmd/main.go:98]: This will overwrite if the state to resume is specified twice. I suppose that's ok?

Comment (floriecai) [reply copydb/cmd/main.go:98]: yeah, state in filepath always wins.

**Pull request merged**